bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitValveImpl.java

```diff
@@ -73,6 +78,73 @@ public HomekitValveImpl(HomekitTaggedItem taggedItem, List<HomekitTaggedItem> ma
         this.activeReader = new BooleanItemReader(getItem(HomekitCharacteristicType.ACTIVE_STATUS, GenericItem.class),
                 OnOffType.ON, OpenClosedType.OPEN);
         getServices().add(new ValveService(this));
+        final String timerConfig = getAccessoryConfiguration(CONFIG_TIMER, "");
+        homekitTimer = timerConfig.equalsIgnoreCase("yes") || timerConfig.equalsIgnoreCase("true");
+        if (homekitTimer) {
+            addRemainingDurationCharacteristic(taggedItem, updater);
+        }
+    }
+
+    private void addRemainingDurationCharacteristic(HomekitTaggedItem taggedItem, HomekitAccessoryUpdater updater) {
+        logger.trace("addRemainingDurationCharacteristic for {}", taggedItem);
+        ((ValveService) getPrimaryService()).addOptionalCharacteristic(new RemainingDurationCharacteristic(() -> {
+            int remainingTime = 0;
+            ScheduledFuture<?> future = valveTimer;
+            if (future != null && !future.isDone()) {
+                remainingTime = java.lang.Math.toIntExact(future.getDelay(TimeUnit.SECONDS));
+            }
+            return CompletableFuture.completedFuture(remainingTime);
+        }, (callback) -> updater.subscribe((GenericItem) taggedItem.getItem(), REMAINING_DURATION.getTag(), callback),
+                () -> updater.unsubscribe((GenericItem) taggedItem.getItem(), REMAINING_DURATION.getTag())
+
+        ));
+    }
+
+    /**
+     * return duration set by home app at corresponding OH items. if ot set, then return the default duration from
+     * configuration.
+     * 
+     * @return duraion
+     */
+    private int getDuration() {
+        int duration = 0;
+        final @Nullable DecimalType durationState = getStateAs(HomekitCharacteristicType.DURATION, DecimalType.class);
+        if (durationState != null) {
+            duration = durationState.intValue();
+        }
+        return duration;
+    }
+
+    private void startTimer() {
+        int duration = getDuration();
+        logger.trace("start timer for duration {}", duration);
+        if (duration > 0) {
+            ScheduledFuture<?> future = valveTimer;
+            if (future != null && !future.isDone()) {
+                future.cancel(true);
+            }
+            valveTimer = timerService.schedule(() -> {
+                logger.trace("valve timer is over. switching off the valve");
+                switchOffValve();
+                // let home app refresh the remaining duration, which is 0
+                ((GenericItem) getRootAccessory().getItem()).send(RefreshType.REFRESH);
+            }, duration, TimeUnit.SECONDS);
+
+            // let home app refresh the remaining duration, which is 0
+            ((GenericItem) getRootAccessory().getItem()).send(RefreshType.REFRESH);
```

Comment:
This should be equivalent unless you really need the valveTimer to be assigned before the refresh command is sent.
```suggestion
            stopTimer();
            valveTimer = timerService.schedule(() -> {
                logger.trace("valve timer is over. switching off the valve");
                switchOffValve();
                // let home app refresh the remaining duration, which is 0
                ((GenericItem) getRootAccessory().getItem()).send(RefreshType.REFRESH);
            }, duration, TimeUnit.SECONDS);
```

Reply:
good one. 
i will also move REFRESH to the caller method so that we have it once, after the timer is set or stopped.